exporters/prometheus/exporter.go

```diff
@@ -381,16 +379,64 @@ func sanitizeName(n string) string {
 	return b.String()
 }
 
-func validateMetrics(name, description string, metricType *dto.MetricType, mfs map[string]*dto.MetricFamily) (drop bool, help string) {
-	emf, exist := mfs[name]
+func (c *collector) metricTypeAndName(m metricdata.Metrics) (*dto.MetricType, string) {
+	name := c.getName(m)
+
+	switch v := m.Data.(type) {
+	case metricdata.Histogram[int64], metricdata.Histogram[float64]:
+		return dto.MetricType_HISTOGRAM.Enum(), name
+	case metricdata.Sum[float64]:
+		if v.IsMonotonic {
+			return dto.MetricType_COUNTER.Enum(), name + counterSuffix
+		}
+		return dto.MetricType_GAUGE.Enum(), name
+	case metricdata.Sum[int64]:
+		if v.IsMonotonic {
+			return dto.MetricType_COUNTER.Enum(), name + counterSuffix
+		}
+		return dto.MetricType_GAUGE.Enum(), name
+	case metricdata.Gauge[int64], metricdata.Gauge[float64]:
+		return dto.MetricType_GAUGE.Enum(), name
+	}
+
+	return nil, ""
+}
+
+func (c *collector) scopeInfo(scope instrumentation.Scope) (prometheus.Metric, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	scopeInfo, ok := c.scopeInfos[scope]
+	if ok {
+		return scopeInfo, nil
+	}
+
+	scopeInfo, err := createScopeInfoMetric(scope)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create scope info metric: %w", err)
+	}
+
+	c.scopeInfos[scope] = scopeInfo
```

Comment:
_Not directly related to this PR_

If non-nil error is returned by `createScopeInfoMetric` then it will be called during each `Collect` for given `scope`.

Reply:
Do you want to address it in this issue as well or shall I create an issue for it?

Reply:
I created https://github.com/open-telemetry/opentelemetry-go/issues/4091

Feel free to address it in a separate PR 😉